src/nodeMatchers.ts

```diff
@@ -1,200 +1,35 @@
+import { findNodeOfType } from "./nodeFinders";
+
+export function matcher(
+  finder: NodeFinder,
+  selector: SelectionExtractor = simpleSelectionExtractor
 ): NodeMatcher {
-  return (editor: TextEditor, node: SyntaxNode) => {
-    const returnNode = getMatchingChildNode(node);
-
-    if (returnNode == null) {
-      return null;
-    }
-
-    return extractor(returnNode);
+  return function (editor: TextEditor, node: SyntaxNode) {
+    const targetNode = finder(node);
+    return targetNode ? selector(editor, targetNode) : null;
   };
 }
 
-export function getNodeWithLeadingDelimiter(
-  node: SyntaxNode
-): SelectionWithContext {
-  const leadingDelimiterToken = node.previousSibling!;
-
-  const leadingDelimiterRange = makeRange(
-    leadingDelimiterToken.startPosition,
-    node.startPosition
-  );
-
-  return {
-    ...simpleSelectionExtractor(node),
-    context: {
-      leadingDelimiterRange,
-    },
-  };
-}
-
-export const notSupported: NodeMatcher = (
-  editor: TextEditor,
-  node: SyntaxNode
-) => {
-  throw new Error("Node type not supported");
-};
-
-function getNextNonDelimiterNode(
-  startNode: SyntaxNode,
-  isDelimiterNode: (node: SyntaxNode) => boolean
-): SyntaxNode | null {
-  var node = startNode.nextSibling;
-
-  while (node != null) {
-    if (!isDelimiterNode(node)) {
-      return node;
-    }
-
-    node = node.nextSibling;
-  }
-
-  return node;
-}
-
-function getPreviousNonDelimiterNode(
-  startNode: SyntaxNode,
-  isDelimiterNode: (node: SyntaxNode) => boolean
-): SyntaxNode | null {
-  var node = startNode.previousSibling;
-
-  while (node != null) {
-    if (!isDelimiterNode(node)) {
-      return node;
-    }
-
-    node = node.previousSibling;
-  }
-
-  return node;
-}
-
-export function delimitedMatcher(
-  nodeMatches: (node: SyntaxNode) => boolean,
-  isDelimiterNode: (node: SyntaxNode) => boolean,
-  defaultDelimiter: string
+export function composedMatcher(
+  finders: NodeFinder[],
+  selector: SelectionExtractor = simpleSelectionExtractor
 ): NodeMatcher {
-  return (editor: TextEditor, node: SyntaxNode) => {
-    if (!nodeMatches(node)) {
-      return null;
-    }
-
-    var containingListDelimiter: string | null = null;
-    var leadingDelimiterRange: Range | null = null;
-    var trailingDelimiterRange: Range | null = null;
-
-    const nextNonDelimiterNode = getNextNonDelimiterNode(node, isDelimiterNode);
-    const previousNonDelimiterNode = getPreviousNonDelimiterNode(
-      node,
-      isDelimiterNode
-    );
-
-    if (nextNonDelimiterNode != null) {
-      trailingDelimiterRange = makeRange(
-        node.endPosition,
-        nextNonDelimiterNode.startPosition
-      );
-
-      containingListDelimiter = editor.document.getText(trailingDelimiterRange);
+  return function (editor: TextEditor, initialNode: SyntaxNode) {
+    let returnNode: SyntaxNode | null = initialNode;
+    for (const finder of finders) {
+      returnNode = returnNode ? finder(returnNode) : null;
```

Comment:
should we not short-circuit if one of the finders returns null?

Reply:
I think it makes more sense to short circuit. Otherwise you could be effectively skipping a finder, which could yield the incorrect node. 

For instance, take this completely made up example. We want to get the return type of a function:

```ts
returnType: composedMatcher([
    findNodeOfType("function_declaration"),
    findNodeOfType("return_type"),
    getTypeNode
])
```

Imagine a situation where we find a declaration but we don't find anything for `findNodeOfType("return_type")`. If we continue to the next finder, we could match any other encountered type (like a parameter type) in the function declaration, returning an incorrect node.

Reply:
ah oh right you *are* short-circuiting.  Missed that.  I think maybe would be a bit easier to understand if you just had a `return` statement when you get a `null`?  Then it's really obvious you're short-cirtuiting and can remove ternary at bottom